File: Android-App/app/src/main/java/com/cleaningmanager/app/data/models/Shift.kt
package com.cleaningmanager.app.data.models

import com.google.gson.annotations.SerializedName

data class Shift(
    @SerializedName("_id")
    val id: String,
    val apartment: ApartmentInfo,
    val cleaner: CleanerInfo,
    val scheduledDate: String,
    val scheduledStartTime: String,
    val scheduledEndTime: String? = null,
    val actualStartTime: String? = null,
    val actualEndTime: String? = null,
    val status: ShiftStatus,
    val notes: String? = null,
    val confirmedSeen: ConfirmedSeen? = null,
    val timeChangeRequest: TimeChangeRequest? = null
)

enum class ShiftStatus {
    scheduled,
    @SerializedName("in_progress")
    in_progress,
    completed,
    cancelled
}

data class ApartmentInfo(
    @SerializedName("_id")
    val id: String,
    val name: String,
    val address: String
)

data class CleanerInfo(
    @SerializedName("_id")
    val id: String,
    val name: String,
    val email: String
)

data class ConfirmedSeen(
    val confirmed: Boolean,
    val confirmedAt: String? = null
)

data class TimeChangeRequest(
    val requestedBy: String,
    val newStartTime: String? = null,
    val newEndTime: String? = null,
    val reason: String? = null,
    val status: String,
    val operatorConfirmed: Boolean? = null
)

data class ShiftsResponse(
    val shifts: List<Shift>
)

data class ShiftResponse(
    val shift: Shift
)

data class CreateShiftRequest(
    val apartment: String,
    val cleaner: String,
    val scheduledDate: String,
    val scheduledStartTime: String,
    val scheduledEndTime: String? = null,
    val notes: String? = null
)

File: iOS-App/CleaningShiftManager/Models/Shift.swift
//
//  Shift.swift
//  CleaningShiftManager
//

import Foundation

struct Shift: Codable, Identifiable {
    let id: String
    let apartment: ApartmentInfo
    let cleaner: CleanerInfo
    let scheduledDate: String
    let scheduledStartTime: String
    let scheduledEndTime: String?
    let actualStartTime: String?
    let actualEndTime: String?
    let status: ShiftStatus
    let notes: String?
    let confirmedSeen: ConfirmedSeen?
    let timeChangeRequest: TimeChangeRequest?
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case apartment
        case cleaner
        case scheduledDate
        case scheduledStartTime
        case scheduledEndTime
        case actualStartTime
        case actualEndTime
        case status
        case notes
        case confirmedSeen
        case timeChangeRequest
    }
    
    enum ShiftStatus: String, Codable {
        case scheduled
        case in_progress
        case completed
        case cancelled
    }
    
    struct ApartmentInfo: Codable {
        let id: String
        let name: String
        let address: String
        
        enum CodingKeys: String, CodingKey {
            case id = "_id"
            case name
            case address
        }
        
        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            
            // Handle _id as string or object
            if let idString = try? container.decode(String.self, forKey: .id) {
                self.id = idString
            } else {
                let idObj = try container.decode(IDWrapper.self, forKey: .id)
                self.id = idObj._id
            }
            
            self.name = try container.decode(String.self, forKey: .name)
            self.address = try container.decode(String.self, forKey: .address)
        }
        
        struct IDWrapper: Codable {
            let _id: String
        }
    }
    
    struct CleanerInfo: Codable {
        let id: String
        let name: String
        let email: String
        
        enum CodingKeys: String, CodingKey {
            case id = "_id"
            case name
            case email
        }
        
        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            
            // Handle _id as string or object
            if let idString = try? container.decode(String.self, forKey: .id) {
                self.id = idString
            } else {
                let idObj = try container.decode(IDWrapper.self, forKey: .id)
                self.id = idObj._id
            }
            
            self.name = try container.decode(String.self, forKey: .name)
            self.email = try container.decode(String.self, forKey: .email)
        }
        
        struct IDWrapper: Codable {
            let _id: String
        }
    }
    
    struct ConfirmedSeen: Codable {
        let confirmed: Bool
        let confirmedAt: String?
    }
    
    struct TimeChangeRequest: Codable {
        let requestedBy: String
        let newStartTime: String?
        let newEndTime: String?
        let reason: String?
        let status: String
        let operatorConfirmed: Bool?
    }
}

struct ShiftsResponse: Codable {
    let shifts: [Shift]
}

struct CreateShiftRequest: Codable {
    let apartment: String
    let cleaner: String
    let scheduledDate: String
    let scheduledStartTime: String
    let scheduledEndTime: String?
    let notes: String?
}

